Samples/ExtendedControlAndMetadata/EyeGazeAndBackgroundSegmentation/CameraKsPropertyHelper/ResourceLock.h

```diff
@@ -0,0 +1,92 @@
+// Copyright (C) Microsoft Corporation
+#pragma once
+#include <Windows.h>
+
+//
+//  Simple CRITICAL_SECTION based locks
+//
+class CResource_Lock
+{
+    // make copy constructor and assignment operator inaccessible
+
+    CResource_Lock(const CResource_Lock& refCritSec);
+    CResource_Lock& operator=(const CResource_Lock& refCritSec);
+
+    CRITICAL_SECTION m_CritSec;
+
+public:
+    CResource_Lock()
+    {
+        InitializeCriticalSection(&m_CritSec);
+    };
+
+    ~CResource_Lock()
+    {
+        DeleteCriticalSection(&m_CritSec);
+    };
+
+    void Lock()
+    {
+        EnterCriticalSection(&m_CritSec);
+    };
+    void Unlock()
+    {
+        LeaveCriticalSection(&m_CritSec);
+    };
+    void LockExclusive()
+    {
+        EnterCriticalSection(&m_CritSec);
+    };
+    void UnlockExclusive()
+    {
+        LeaveCriticalSection(&m_CritSec);
+    };
+    bool IsLockHeldByCurrentThread()
+    {
+        return GetCurrentThreadId() == static_cast<DWORD>(reinterpret_cast<ULONG_PTR>(m_CritSec.OwningThread));
+    };
+    bool IsLockHeld()
+    {
+        return m_CritSec.OwningThread != 0;
+    };
+    bool TryLock()
+    {
+        return TryEnterCriticalSection(&m_CritSec);
+    };
+    // aliased methods to help code compat so that CriticalSections can be passed to ReaderWriter templates
+    void LockShared()
+    {
+        EnterCriticalSection(&m_CritSec);
+    };
+    void UnlockShared()
+    {
+        LeaveCriticalSection(&m_CritSec);
+    };
+};
+
+// locks a critical section, and unlocks it automatically
+// when the lock goes out of scope
+class CResource_AutoLock

```

Comment:
>CResource_AutoLock [](http://example.com/codeflow?start=6&length=18)

For open source project I would use std::mutex or std::recursive_mutex

Reply:
yeah if I was using other std commodities or mixing another framework I would. It is ok to show available windows.h provided sync mechanism

Reply:
I just thought it would be more standard coding practice and not as much code ramp up for a new dev